# Disallow characters which are made with multiple code points in character class syntax (no-misleading-character-class)

Unicode includes the characters which are made with multiple code points.
RegExp character class syntax (`/[abc]/`) cannot handle characters which are made by multiple code points as a character; those characters will be dissolved to each code point. For example, `â‡ï¸` is made by `â‡` (`U+2747`) and VARIATION SELECTOR-16 (`U+FE0F`). If this character is in RegExp character class, it will match to either `â‡` (`U+2747`) or VARIATION SELECTOR-16 (`U+FE0F`) rather than `â‡ï¸`.

This rule reports the regular expressions which include multiple code point characters in character class syntax. This rule considers the following characters as multiple code point characters.

**A character with combining characters:**

The combining characters are characters which belong to one of `Mc`, `Me`, and `Mn` [Unicode general categories](http://www.unicode.org/L2/L1999/UnicodeData.html#General%20Category).

```js
/^[AÌ]$/u.test("AÌ") //â†’ false
/^[â‡ï¸]$/u.test("â‡ï¸") //â†’ false
```

**A character with Emoji modifiers:**

```js
/^[ğŸ‘¶ğŸ»]$/u.test("ğŸ‘¶ğŸ»") //â†’ false
/^[ğŸ‘¶ğŸ½]$/u.test("ğŸ‘¶ğŸ½") //â†’ false
```

**A pair of regional indicator symbols:**

```js
/^[ğŸ‡¯ğŸ‡µ]$/u.test("ğŸ‡¯ğŸ‡µ") //â†’ false
```

**Characters that ZWJ joins:**

```js
/^[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]$/u.test("ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦") //â†’ false
```

**A surrogate pair without Unicode flag:**

```js
/^[ğŸ‘]$/.test("ğŸ‘") //â†’ false

// Surrogate pair is OK if with u flag.
/^[ğŸ‘]$/u.test("ğŸ‘") //â†’ true
```

## Rule Details

This rule reports the regular expressions which include multiple code point characters in character class syntax.

Examples of **incorrect** code for this rule:

```js
/*eslint no-misleading-character-class: error */

/^[AÌ]$/u
/^[â‡ï¸]$/u
/^[ğŸ‘¶ğŸ»]$/u
/^[ğŸ‡¯ğŸ‡µ]$/u
/^[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]$/u
/^[ğŸ‘]$/
```

Examples of **correct** code for this rule:

```js
/*eslint no-misleading-character-class: error */

/^[abc]$/
/^[ğŸ‘]$/u
```

## When Not To Use It

You can turn this rule off if you don't want to check RegExp character class syntax for multiple code point characters.
