---
title: no-await-in-loop
rule_type: problem
---


Performing an operation on each element of an iterable is a common task. However, performing an
`await` as part of each operation is an indication that the program is not taking full advantage of
the parallelization benefits of `async`/`await`.

Usually, the code should be refactored to create all the promises at once, then get access to the
results using `Promise.all()`. Otherwise, each successive operation will not start until the
previous one has completed.

Concretely, the following function should be refactored as shown:

```js
async function foo(things) {
  const results = [];
  for (const thing of things) {
    // Bad: each loop iteration is delayed until the entire asynchronous operation completes
    results.push(await bar(thing));
  }
  return baz(results);
}
```

```js
async function foo(things) {
  const results = [];
  for (const thing of things) {
    // Good: all asynchronous operations are immediately started.
    results.push(bar(thing));
  }
  // Now that all the asynchronous operations are running, here we wait until they all complete.
  return baz(await Promise.all(results));
}
```

Also, suppose a loop iterates over a collection of `Promise` rejections, that already exists, with `await` in that case, after the first promise is rejected, it immediately throws an error, stopping the loop's execution before handling the rest of the promises. This leads to unhandled rejections for promises that havenâ€™t been awaited yet.

```js
async function bar(things) {
  const arrayOfRejections = new Array(things).fill(undefined).map((_, index) => Promise.reject(new Error("rejected " + index)));
  const result = [];
  for (const promise of arrayOfRejections) {
    // Bad: after the rejection of first promise, loop's execution stops, leaving the rest of the promise rejections floating and unhandled.
    const promisedValue = await promise;
    result.push(promisedValue);
  }

  return result;
}
```

This function should be refactored as shown:

```js
async function bar(things) {
  const arrayOfRejections = new Array(things).fill(undefined).map((_, index) => Promise.reject(new Error("rejected " + index)));
  const result = [];
  // Good: if any of the promises rejects, the Promise.all() call will reject, and any rejections in the other promises will not be treated as unhanlded rejections.
  for (const promisedValue of await Promise.all(arrayOfRejections)) {
    result.push(promisedValue);
  }

  return result;
}
```

## Rule Details

This rule disallows the use of `await` within loop bodies.

## Examples

Examples of **correct** code for this rule:

:::correct

```js
/*eslint no-await-in-loop: "error"*/

async function foo(things) {
  const results = [];
  for (const thing of things) {
    // Good: all asynchronous operations are immediately started.
    results.push(bar(thing));
  }
  // Now that all the asynchronous operations are running, here we wait until they all complete.
  return baz(await Promise.all(results));
}
```

:::

Examples of **incorrect** code for this rule:

:::incorrect

```js
/*eslint no-await-in-loop: "error"*/

async function foo(things) {
  const results = [];
  for (const thing of things) {
    // Bad: each loop iteration is delayed until the entire asynchronous operation completes
    results.push(await bar(thing));
  }
  return baz(results);
}
```

:::

## When Not To Use It

In many cases the iterations of a loop are not actually independent of each-other. For example, the
output of one iteration might be used as the input to another. Or, loops may be used to retry
asynchronous operations that were unsuccessful. Or, loops may be used to prevent your code from sending
an excessive amount of requests in parallel. Also, if `Promise` rejections are generated by `function generators`, then it's possible that iterate those rejections with `await` through loop may well be correct. such as:

```js
function * yieldPromises() {
  let n = 1;
  try {
    while (true) {
      if (n > 10) {
        yield Promise.reject(new Error(`Promise ${n}`));
      } else {
        yield Promise.resolve(`Promise ${n}`);
      }
      n++;
    }
  } finally {
    console.log('promise generator closed');
  }
}

async function awaitUntilRejection() {
  for (const promise of yieldPromises()) {
    try {
      const fulfillmentValue = await promise;
      console.log(`Fulfillment value: ${fulfillmentValue}`);
    } catch (e) {
      console.log(`first rejection message: ${e}`);
      break;
    }
  }
}

// no unhandled rejections, since no other promises were even created.
awaitUntilRejection();
```

In such cases it makes sense to use `await` within a
loop and it is recommended to disable the rule via a standard ESLint disable comment.
